Dockerfile

```diff
@@ -1,56 +1,26 @@
-        bats jq podman runc \
-	golint \
-	openssl \
-	&& dnf clean all
-
-# Install two versions of the registry. The first is an older version that
-# only supports schema1 manifests. The second is a newer version that supports
-# both. This allows integration-cli tests to cover push/pull with both schema1
-# and schema2 manifests.
-RUN set -x \
-	&& export GO111MODULE=off \
-	&& REGISTRY_COMMIT_SCHEMA1=ec87e9b6971d831f0eff752ddb54fb64693e51cd \
-	&& REGISTRY_COMMIT=47a064d4195a9b56133891bbb13620c3ac83a827 \
-	&& export GOPATH="$(mktemp -d)" \
-	&& git clone https://github.com/docker/distribution.git "$GOPATH/src/github.com/docker/distribution" \
-	&& (cd "$GOPATH/src/github.com/docker/distribution" && git checkout -q "$REGISTRY_COMMIT") \
-	&& GOPATH="$GOPATH/src/github.com/docker/distribution/Godeps/_workspace:$GOPATH" \
-		go build -o /usr/local/bin/registry-v2 github.com/docker/distribution/cmd/registry \
-	&& (cd "$GOPATH/src/github.com/docker/distribution" && git checkout -q "$REGISTRY_COMMIT_SCHEMA1") \
-	&& GOPATH="$GOPATH/src/github.com/docker/distribution/Godeps/_workspace:$GOPATH" \
-		go build -o /usr/local/bin/registry-v2-schema1 github.com/docker/distribution/cmd/registry \
-	&& rm -rf "$GOPATH"
+ENV GOPATH="/var/tmp/go"
+ENV GOSRC="$GOPATH/src/github.com/containers/skopeo"
+ENV PATH="$GOPATH/bin:$GOSRC/bin:/usr/share/gocode/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin"
 
-RUN set -x \
-	&& export GO111MODULE=off \
-	&& export GOPATH=$(mktemp -d) \
-	&& git clone --depth 1 -b v1.5.0-alpha.3 git://github.com/openshift/origin "$GOPATH/src/github.com/openshift/origin" \
-	# The sed edits out a "go < 1.5" check which works incorrectly with go ≥ 1.10. \
-	&& sed -i -e 's/\[\[ "\${go_version\[2]}" < "go1.5" ]]/false/' "$GOPATH/src/github.com/openshift/origin/hack/common.sh" \
-	&& (cd "$GOPATH/src/github.com/openshift/origin" && make clean build && make all WHAT=cmd/dockerregistry) \
-	&& cp -a "$GOPATH/src/github.com/openshift/origin/_output/local/bin/linux"/*/* /usr/local/bin \
-	&& cp "$GOPATH/src/github.com/openshift/origin/images/dockerregistry/config.yml" /atomic-registry-config.yml \
-	&& rm -rf "$GOPATH" \
-	&& mkdir /registry
-
-ENV GOPATH /usr/share/gocode:/go
-ENV PATH $GOPATH/bin:/usr/share/gocode/bin:$PATH
-ENV container_magic 85531765-346b-4316-bdb8-358e4cca9e5d
-RUN go version
-WORKDIR /go/src/github.com/containers/skopeo
-COPY . /go/src/github.com/containers/skopeo
+RUN dnf -y update \
+	&& dnf -y install \
+		make git golang golang-github-cpuguy83-md2man \
+		btrfs-progs-devel device-mapper-devel \
+		libassuan-devel gpgme-devel gnupg \
+		httpd-tools which tar wget hostname util-linux \
+		bsdtar socat ethtool device-mapper iptables \
+		tree findutils nmap-ncat e2fsprogs xfsprogs lsof \
+		docker iproute bats jq podman runc golint \
+		openssl \
+	&& dnf -y upgrade \
+	&& dnf erase -y skopeo \
+	&& dnf clean all
 
-#ENTRYPOINT ["hack/dind"]
+COPY . $GOSRC
```

Comment:
Having this `COPY` last allows more sharing on interactive debugging of the tests. Maybe start with a `COPY hack`, or even more specific, and end with a `COPY .`?

Reply:
Yes, I noticed that compile/setup time, I don't like it either.  I think probably the best way to handle this for both developers and CI, is to separately build a base-image.  Then use that here to add in only the finishing touches, like overwriting $GOSRC in the container.  That way we gain all the benefits of environment-consistency, it will be reasonably up-to-date (i.e. built daily), and fast for CI + Developer use.

The disadvantages are: A daily out-of-date base image, means both CI and Developers will always be pulling it down - annoying, but quite likely faster than building + compiling each time.  The other disadvantage is it will absolutely be a 'rolling' environment tightly-coupled to work on the *main* branch only.  It will be useless/dangerous trying to use it on any release branches - but maybe this is a separate discussion.

Reply:
In case it sparks an idea:  What we do in podman and buildah is rely on container images built at the same time as VM images.  They 99.999% match the VM setup, and **NEVER** use a `latest` tag (i.e. only `c123456789` strings matching the VM images).  This means the CI and developer experience on both *main* and all release branches remains consistent - it only ever changes when someone deliberately changes the ID.

The down-side here is the container images are VERY large (almost 1gb) with almost zero room for reduction.  On the other hand, they change much more slowly/deliberately compared to a rolling `latest` image.  IMHO this makes development work easier, because you don't have *two* moving targets (the environment and the WIP code).

The other down-side is, it's desirable to keep `.cirrus.yml` as the single source-of-truth for the image ID (and tag).  That complicates operations outside of CI, since there's an extra step in obtaining the *current* ID (e.g. parsing YAML to JSON and filtering the ID with `jq`).  It can be done, it's just an extra step.

Reply:
> I think probably the best way to handle this for both developers and CI, is to separately build a base-image.

That’s very attractive for _most_ cases. IIRC the OpenShift compilation is by far the most resource-intensive part of Skopeo builds, and letting users just download a binary would be convenient.

Looking at the history of `Dockerfile`, it only changes a few times a year, so having a _named_ image pulled down instead of relying on local build cache feels like an improvement. I don’t have a numbers on the (world-wide) relative cost of pulling a gigabyte of data vs. a local OpenShift build, but it seems reasonable to guess that pulling the gigabyte image would not be _prohibitively_ worse for anyone.

_Occasionally_, some changes to the `Dockerfile` happen in bursts of development, and at that time it should be possible to work with a locally-built container; so there’s a downside in having to maintain, or at least enable, a rarely-used code path. (Also, there would need to be a way to trigger a fresh image/VM build after such a change happens.)

WRT using rolling `latest` vs. non-rolling explicitly tagged images/VMs, I don’t have a strong opinion. It sounds important to have some, preferably automated, or at least easy, way to run tests on old stable branches, but you have much better idea what the trade-offs/limitations in this area (I vaguely remember something about expiring images?) are.

Reply:
> pulling the gigabyte image would not be prohibitively worse for anyone.

I know anybody with a poor internet link would absolutely hate it and curse the universe. In fact, random networking glitches have caused Cirrus-CI to timeout running jobs (maybe 10 times in the last few months) due to the image size.  I also don't know how in-sync quay.io's content-distribution proxies are (or if they even exist).  So it could be quite slow for anybody network-far from the source.  Give it a try: `podman pull quay.io/libpod/fedora_podman:c4687311653830656` (chosen at random).

> Also, there would need to be a way to trigger a fresh image/VM build after such a change happens

If I setup a dedicated base-image build for this on `c/automation_images`, needing to hand-build is actually a very easy and relatively un-complicated `make foobar` type command (on that repo).  It's the VM images which are complex.  This will keep the size down as well.

> idea what the trade-offs/limitations in this area

I have an idea of how best to keep automation stable/reliable.  It's the local-developer workflow that I'm much less savvy about.  The two often have conflicting priorities of "latest" vs "stable".  It's possible to solve for both, but adds risk of sometimes CI != Development (in very non-obvious ways).

> I vaguely remember something about expiring images

That's a big problem on the VM side, and a small problem for container images.  I'm not sure of the exact details, but I am sure that quay.io will delete tagged images when they're not pulled for some interval.  However, it sounds to me like I need to go ahead and setup cirrus-cron jobs on skopeo release branches anyway (to preserve VM images).  Doing that will also cause the container images to be pulled, and (in theory) block their removal.

Reply:
To be clear: What I'm thinking of is setting up `c/automation_images` to build a base image which is consumed by the Dockerfile (here).  It will produce a `latest` image daily, and a tagged image every time updated VM images are called for.  I'll add in a `ARG tag=latest`, but have CI override that at runtime with `tag=${IMAGE_SUFFIX}`.

Secondarily, I'll setup cirrus-cron jobs to run on skopeo release branches.  Going-forward (as new release branches are made) this will block removal of both VM and the tagged container base image.  However the "old" `latest` image will obviously be lost so development work on release branches will need a way to override the `tag` arg via a `make` option.

Clear as mud? :grin:

Reply:
> > pulling the gigabyte image would not be prohibitively worse for anyone.
> 
> I know anybody with a poor internet link would absolutely hate it and curse the universe.

Sure, but (assuming the manually-tagged rarely-changing images) the alternative is not magic that pulls down 0 bytes, but all the `git clone`s and RPM updates necessary to build the container image locally, with many more round-trips and other latency than a straight pull of a gigabyte of data.

Reply:
> To be clear: What I'm thinking of is setting up `c/automation_images` to build a base image which is consumed by the Dockerfile (here). It will produce a `latest` image daily, and a tagged image every time updated VM images are called for. I'll add in a `ARG tag=latest`, but have CI override that at runtime with `tag=${IMAGE_SUFFIX}`.

Sounds reasonable. Maybe have the `Makefile` for local builds also use the same tag, so that local users (if any) don’t keep re-downloading every day?

[Anyway is all of this predicated on fixing the OpenShift failure we are seeing?]

Reply:
> not magic that pulls down 0 bytes

Of course not.  The 1g image is generic, and matches the VM images.  I think I can do a custom/special-purpose base-image just for Skopeo that will be smaller, maybe half that size.  So twice as fast to download, at the cost of some extra tinkering in c/image (which I have no problem with doing).

Reply:
> local builds also use the same tag

This would be ideal, as we could completely avoid the whole ever-changing 'latest' image.  The cost is in extra complexity needed to obtain the `$IMAGE_SUFFIX` value out of `.cirrus.yml`, but this is doable.  In fact, I have a small container-image which can do it in a generic way.